src/System.Windows.Forms.Primitives/src/Windows/Win32/System/Com/SafeArrayScope.cs

```diff
@@ -44,70 +63,117 @@ public SafeArrayScope(uint size)
-                }
-
-                return (T)(object)result;
+                return (T)(object)GetElement<int>(i);
+            }
+            else if (typeof(T) == typeof(double))
+            {
+                return (T)(object)GetElement<double>(i);
+            }
+            else if (typeof(T) == typeof(nint))
+            {
+                return (T)(object)GetElement<nint>(i);
             }
 
-            // Noop. We should never get here as we would've thrown in the constructor
-            // for an unknown type.
+            // Noop. This is an unknown type. We should fill this method out to to do the right
+            // thing as we run into new types.
             return default;
         }
         set
         {
-            Span<int> indices = stackalloc int[] { i };
             if (value is string s)
             {
                 using BSTR bstrText = new(s);
-                fixed (int* pIndices = indices)
-                {
-                    PInvoke.SafeArrayPutElement(Value, pIndices, bstrText).ThrowOnFailure();
-                }
+                PutElement(i, bstrText);
             }
             else if (value is int @int)
             {
-                fixed (int* pIndices = indices)
-                {
-                    PInvoke.SafeArrayPutElement(Value, pIndices, &@int).ThrowOnFailure();
-                }
+                PutElement(i, &@int);
             }
+            else if (value is double dbl)
+            {
+                PutElement(i, &dbl);
+            }
+            else if (value is nint @nint)
+            {
+                PutElement(i, (void*)@nint);
+            }
+        }
+    }
+
+    private TReturn GetElement<TReturn>(int index) where TReturn : unmanaged
+    {
+        Span<int> indices = [index];
+        TReturn result;
+        fixed (int* pIndices = indices)
+        {
+            PInvoke.SafeArrayGetElement(Value, pIndices, &result).ThrowOnFailure();
         }
+
+        return result;
     }
 
-    public readonly bool IsNull => Value is null;
+    private void PutElement(int index, void* value)
+    {
+        Span<int> indices = [index];
+        fixed (int* pIndices = indices)
+        {
+            PInvoke.SafeArrayPutElement((SAFEARRAY*)_value, pIndices, value).ThrowOnFailure();
+        }
+    }
+
+    public int Length => (int)Value->GetBounds().cElements;
+
+    public bool IsNull => _value == 0;
+
+    public bool IsEmpty => Length == 0;
 
     public void Dispose()
     {
-        if (!IsNull)
+        if (IsNull)
+        {
+            return;
+        }
+
+        SAFEARRAY* safeArray = (SAFEARRAY*)_value;
+
+        // Really want _value to be null after disposal to avoid double destroy, but we also want
+        // to maintain the readonly state of the struct to allow passing as `in` without creating implicit
+        // copies (which would break the T** and void** operators).
+        *(void**)_value = null;
```

Comment:
This needs to be the `this` pointer.

```suggestion
        *(void**)this = null;
```